### ISSUE TITLE
False negative when using the power (**) operator

### ISSUE THREAD

**Issue by nikola-benes** [python/mypy#7765]
When checking the following function with mypy 0.730 (Python version 3.7.4):

 ```Python
def foo(x: int) -> str:
    return 2 ** x
```

mypy reports no errors:

    Success: no issues found in 1 source file

With the `--strict` option turned on, mypy reports:

    foo.py:2: error: Returning Any from function declared to return "str"

The expected output would be something like

    foo.py:2: error: Incompatible return value type (got "int", expected "str")

**Comment by srittau** [comment]
This belong in [typeshed](/python/typeshed).

This is a limitation in the type system. `foo` can return either an `int` or a `float`, depending on whether `x` is a positive or negative number. While we could change typeshed in a way that `2 ** x` will always return `float` (which is correct, since for typing purposes `int` is a subtype if `float`), this would be rather annoying for users that know that it will return an `int`, for example because they know that `x` is always non-negative. Therefore we return `Any`, because there is no way to properly express this.

We would need a range type to be able to express this.

**Comment by nikola-benes** [comment]
Wouldn't `Union[int, float]` be a better option instead of `Any`? The users can then assert that the result is an int, if they are sure that x is non-negative.

**Comment by nikola-benes** [comment]
Also, a related problem is that

```Python
def foo(x: int) -> float:
    return 2 ** x
```

fails with `mypy --strict`. What is the recomended way to deal with this? Explicit type conversion (`float(2 ** x)`)?

**Comment by msullivan** [comment]
> Wouldn't `Union[int, float]` be a better option instead of `Any`? The users can then assert that the result is an `int` if they are sure that `x` is non-negative.

We try to avoid that sort of union return in typeshed stubs because they produce a pretty bad user experience.

**Comment by msullivan** [comment]
> Also, a related problem is that
> 
> ```python
> def foo(x: int) -> float:
>     return 2 ** x
> ```
> 
> fails with `mypy --strict`. What is the recomended way to deal with this? Explicit type conversion (`float(2 ** x)`)?

Yeah, probably. `--disallow-return-any` maybe shouldn't be in --strict? See https://github.com/python/mypy/issues/7767 which I just opened
